core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala

```diff
@@ -0,0 +1,214 @@
+        
+        Code.get(rc) match {
+          case Code.OK =>
+            // Set ACL for each child
+            if(list != null) 
+              Future {
+                val childPathBuilder = new StringBuilder
+                for(child <- list) {
+                  childPathBuilder.clear
+                  childPathBuilder.append(path)
+                  childPathBuilder.append("/")
+                  childPathBuilder.append(child)
+
+                  val childPath = childPathBuilder.toString
+                  counter.incrementAndGet()
+                  zkHandle.setACL(childPath, ZkUtils.DefaultAcls(true), -1, SetACLCallback, ctx)
+                  zkHandle.getChildren(childPath, false, GetChildrenCallback, ctx)
+                }
+                counter.decrementAndGet()
+              }(threadExecutionContext)
+          case Code.CONNECTIONLOSS =>
+            setAclsRecursively(zkUtils, path)
+          case Code.NONODE =>
+            warn("Node is gone, it could be have been legitimately deleted: %s".format(path))
+          case Code.SESSIONEXPIRED =>
+            // Starting a new session isn't really a problem, but it'd complicate
+            // the logic of the tool, so we quit and let the user re-run it.
+            error("ZooKeeper session expired while changing ACLs")
+            exception = throw ZkException.create(KeeperException.create(rc))
+          case _ =>
+            error("Unexpected return code: %d".format(rc))
+            exception = throw ZkException.create(KeeperException.create(rc))
+        }
+      }
+  }
+  
+  object SetACLCallback extends StatCallback {
+    def processResult(rc: Int,
+                      path: String,
+                      ctx: Object,
+                      stat: Stat) {
+      val zkUtils: ZkUtils = ctx.asInstanceOf[ZkUtils]
+      val zkHandle = zkUtils.zkConnection.getZookeeper
+        
+      Code.get(rc) match {
+          case Code.OK =>
+            info("Successfully set ACLs for %s".format(path))
+            counter.decrementAndGet()
+          case Code.CONNECTIONLOSS =>
+            Future {
+              zkHandle.setACL(path, ZkUtils.DefaultAcls(true), -1, SetACLCallback, ctx)
+            }(threadExecutionContext)
+          case Code.NONODE =>
+            warn("Node is gone, it could be have been legitimately deleted: %s".format(path))
+          case Code.SESSIONEXPIRED =>
+            // Starting a new session isn't really a problem, but it'd complicate
+            // the logic of the tool, so we quit and let the user re-run it.
+            error("ZooKeeper session expired while changing ACLs")
+            exception = throw ZkException.create(KeeperException.create(Code.get(rc)))
+          case _ =>
+            error("Unexpected return code: %d".format(rc))
+            exception = throw ZkException.create(KeeperException.create(Code.get(rc)))   
+      }
+    }
+  }
+  
+  def outputUsage(parser: OptionParser) {
```

Comment:
`def outputUsage(parser: OptionParser): Nothing = {`

The scala compiler will know that this method does not return normally (ie it always throws an exception).